File: matlab/src/opt/+opt/ActivityThinkTime.m
classdef ActivityThinkTime < opt.DecisionVariable
    % ActivityThinkTime  Optimize the activity-level think time of an LQN
    % Activity (continuous). Mirrors native-Python ActivityThinkTime.

    properties
        activity
        minValue
        maxValue
    end

    methods
        function obj = ActivityThinkTime(activity, bounds, name)
            actName = opt.Layered.elemName(activity);
            if nargin < 3 || isempty(name)
                name = [actName '_thinktime'];
            end
            obj@opt.DecisionVariable(name);
            obj.activity = actName;
            obj.minValue = bounds(1);
            obj.maxValue = bounds(2);
        end

        function b = getBounds(obj), b = opt.DecisionVariable.unitBounds(1); end

        function v = decode(obj, x)
            v = obj.minValue + x(1) * (obj.maxValue - obj.minValue);
        end

        function apply(obj, model, value)
            act = opt.Layered.resolveActivity(model, obj.activity);
            if ~isempty(act)
                act.setThinkTime(double(value));
            end
        end

        function t = getVariableType(obj), t = 'think_time'; end

        function layers = getLayer(obj, model)
            layers = {};
            task = opt.Layered.taskOfActivity(model, obj.activity);
            if ~isempty(task), layers{end+1} = task.getName(); end
            proc = opt.Layered.activityProcessorName(model, obj.activity);
            if ~isempty(proc), layers{end+1} = proc; end
            if isempty(layers), layers = {}; end
        end

        function v = currentValue(obj, model)
            v = [];
            act = opt.Layered.resolveActivity(model, obj.activity);
            if ~isempty(act)
                v = opt.Layered.distMean(act.thinkTime);
            end
        end
    end
end

File: jar/src/main/java/jline/opt/variables/ActivityThinkTime.java
package jline.opt.variables;

import jline.lang.layered.Activity;
import jline.lang.layered.LayeredNetwork;
import jline.lang.layered.Task;
import jline.opt.LqnAdapter;

import java.util.ArrayList;
import java.util.List;

/**
 * Optimize the activity-level think time of an LQN Activity (continuous).
 * Mirrors native-Python {@code ActivityThinkTime}.
 */
public class ActivityThinkTime extends LqnDecisionVariable {

    private final String activity;
    private final double minValue;
    private final double maxValue;

    public ActivityThinkTime(Activity activity, double minValue, double maxValue) {
        this(activity.getName(), minValue, maxValue, activity.getName() + "_thinktime");
    }

    public ActivityThinkTime(String activityName, double minValue, double maxValue) {
        this(activityName, minValue, maxValue, activityName + "_thinktime");
    }

    public ActivityThinkTime(String activityName, double minValue, double maxValue, String name) {
        super(name);
        this.activity = activityName;
        this.minValue = minValue;
        this.maxValue = maxValue;
    }

    public double[][] getBounds() {
        return unitBounds(1);
    }

    public Object decode(double[] x) {
        return minValue + x[0] * (maxValue - minValue);
    }

    public void applyLqn(LayeredNetwork model, Object value) {
        Activity act = LqnAdapter.resolveActivity(model, activity);
        if (act != null) {
            act.setThinkTime(((Number) value).doubleValue());
        }
    }

    public String getVariableType() {
        return "think_time";
    }

    public List<String> getLayer(Object model) {
        LayeredNetwork lqn = (LayeredNetwork) model;
        Activity act = LqnAdapter.resolveActivity(lqn, activity);
        Task task = (act != null) ? act.getParent() : null;
        String proc = LqnAdapter.activityProcessorName(lqn, activity);
        List<String> layers = new ArrayList<String>();
        if (task != null) {
            layers.add(task.getName());
        }
        if (proc != null) {
            layers.add(proc);
        }
        return layers.isEmpty() ? null : layers;
    }

    public Object currentValue(Object model) {
        Activity act = LqnAdapter.resolveActivity((LayeredNetwork) model, activity);
        return (act != null) ? Double.valueOf(act.getThinkTimeMean()) : null;
    }
}
